rllib/algorithms/dt/dt_torch_policy.py

```diff
@@ -0,0 +1,553 @@
+                    TIMESTEPS: [batch_size, max_seq_len - 1],
+                }
+            state_batches: RNN states (not used).
+            kwargs: forward compatibility (unused) kwargs.
+
+        Returns:
+            A tuple of: (
+                batched input to the action distribution clas,
+                the action distribution class,
+                updated RNN state (unused),
+            )
+        """
+        # Note: this doesn't create a new SampleBatch, so changes to obs_batch persists
+        obs_batch = self._lazy_tensor_dict(obs_batch)
+
+        batch_size = obs_batch[SampleBatch.OBS].shape[0]
+
+        # NOTE: This is probably the most confusing part of the code, made to work with
+        # env_runner and SimpleListCollector during evaluation, and thus should
+        # be changed for the new Policy and Connector API.
+        # So I'll explain how it works.
+
+        # Add current timestep (+1 because -1 is first observation)
+        # NOTE: ViewRequirement of timestep is -(max_seq_len-2):0.
+        # The wierd limits is because RLlib treats initial obs as time -1,
+        # and then 0 is (act, rew, next_obs), etc.
+        # So we only collect max_seq_len-1 from the rollout and create the current
+        # step here by adding 1.
+        # Decision transformer treats initial observation as timestep 0, giving us
+        # 0 is (obs, act, rew).
+        timesteps = obs_batch[SampleBatch.T]
+        new_timestep = timesteps[:, -1:] + 1
+        obs_batch[SampleBatch.T] = torch.cat([timesteps, new_timestep], dim=1)
+
+        # mask out any padded value at start of rollout
+        # NOTE: the other reason for doing this is that evaluation rollout front
+        # pads timesteps with -1, so using this we can find out when we need to mask
+        # out the front section of the batch.
+        obs_batch[SampleBatch.ATTENTION_MASKS] = torch.where(
+            obs_batch[SampleBatch.T] >= 0, 1.0, 0.0
+        )
+
+        # Remove out-of-bound -1 timesteps after attention mask is calculated
+        uncliped_timesteps = obs_batch[SampleBatch.T]
+        obs_batch[SampleBatch.T] = torch.where(
+            uncliped_timesteps < 0,
+            torch.zeros_like(uncliped_timesteps),
+            uncliped_timesteps,
+        )
+
+        # Computes returns-to-go.
+        # NOTE: There are two rtg calculations: updated_rtg and initial_rtg.
+        # updated_rtg takes the previous rtg value (the ViewRequirement is
+        # -(max_seq_len-1):-1), and subtracts the last reward from it.
+        rtg = obs_batch[SampleBatch.RETURNS_TO_GO]
+        last_rtg = rtg[:, -1]
+        last_reward = obs_batch[SampleBatch.REWARDS]
+        updated_rtg = last_rtg - last_reward
+        # initial_rtg simply is filled with target_return.
+        # These two are both only for the current timestep.
+        initial_rtg = torch.full(
+            (batch_size, 1),
+            fill_value=self.config["target_return"],
```

Comment:
can we create a method in this policy `get_initial_return(shape, device, dtype)` which returns a tensor of correct shape for the the initial rtg? By default the implementation will be what you have now if target_return is specified in config. Otherwise raise an error. This way user can implement their own logic later (say they want to evaluate their policy across a range of rtg values? not just a single one.) They can then override the method to return a random set of rtg values.

Reply:
added